### ISSUE TITLE
Slow typing in large documents

### ISSUE THREAD

**Issue by waterfallpit** [Zettlr/Zettlr#3469]
### Description

I noticed slow typing (a certain lag with which the characters appear in the document) for large documents (e.g., ~200 headings, ~10 000 words). After recording with the DevTools I saw that TOC and document info (word count) are updated for every keypress. This is especially noticable on slower computers or when running other somewhat CPU-demandings apps (say, videoconference apps).

### Proposed Changes

Updating TOC and document info after a delay via a timer would speed up the typing with only a short delay for TOC and the document info update.

I implemented this by setting a 400 ms timer for `updateTableOfContents` when fired due to the editor's `change` event. Setting such a timer for word count would speed thing up even more. However, if I remove committing `activeDocumentInfo` from `cursorActivity` event, then the cursor position is not updated correctly.

### Caveats

TOC and document info are updated with a short (<1 s) delay.

### Do you Wish to Attempt Implementing this Yourself?

Yes

### Zettlr Version

Right off the develop branch

### Your Platform

- [X] Windows
- [ ] macOS
- [ ] Linux

### Operating System Version

Windows 10

**Comment by nathanlesage** [comment]
I habe actually implemented such a delay once, but removed it because it didn't feel snappy anymore and it became hugely annoying.

What about instead of utilizing a timeout, trying an idle frame? This way updating would be instantaneous if your computer is fast enough but preference would still be given to typing?

**Comment by waterfallpit** [comment]
It felt snappy to me, but that's probably because I never look at the TOC or the word count or the position indicator when I'm typing.
I don't know how to implement it by waiting for the idle period other than to rely on the timer. Do you have something specific in mind?

**Comment by nathanlesage** [comment]
I was thinking about [this](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback).

**Comment by waterfallpit** [comment]
Thank you for the suggestion. I rewrote the code and tested both versions (with timeout and with idle callback) while running prime95. Both feel snappier than without a callback. I'll test them some more on a slower computer and report back. My concern is that requestIdleCallback often executes the callback right after a keypress and then the following keypress could be slowed down, since the callback is still executing and the new keypress does not cancel the callback once that one is called.

**Comment by waterfallpit** [comment]
I was testing it today and requestIdleCallback does make a positive difference in responsiveness.
But then I was poking around the `updateTableOfContents` to see how to reduce the amount of processing of headings, since most of the changes to the text via typing do not change the headings and therefore the TOC. The following code there:
```
        for (const entry of toc) {
          entry.text = md2html(entry.text)
          entry.text = sanitizeHtml(entry.text, {
            // Headings may be emphasised and contain code
            allowedTags: [ 'em', 'kbd', 'code' ]
          })
        }
```
could be optimized to process the `entry.text` only if it has changed. There is, however, a bug: `<em>...</em>` tags in the headings are interpreted as italic text. Simple swiching of the function calls and settings `allowedTags: [ '' ]` solves this bug, but then `md2html()` introduces `<p>...</p>` into the headings, which are not removed. Running `sanitizeHtml()` again seems wasteful, but if it is run only on headings containing markup, it would probably still require less time on the whole.

I'll test some more and report back.

**Comment by nathanlesage** [comment]
With regard to reparsing the Markdown: I also suspect that the biggest performance hit may well come from sanitizeHTML, therefore, a simple memoize-function within the scope of that Vue component may already be sufficient.

**Comment by waterfallpit** [comment]
OK, so I replaced `sanitizeHtml()` with simpler code for processing text in the headings for the TOC and added some ifs:
- if the table of contents is empty (`null`), then all headings' text is processed,
- if the length of the current TOC differs from the one generated from the file, then, likewise,  all headings' texts are processed,
- else, only the headings, where the text in TOC differs from the text in a specific heading, are processed (e.g., a user is typing a heading),
- since markup processing changes the text of the heading that is to be added to the TOC, all headings' texts with markup are processed, even if the user has not changed them.

Two tests done on my few-year-old laptop with an i5 CPU:
Test file 1: ~75000 words, 1000 headings, 100 of them with markup:
Typing if sufficiently fast, `updateTableOfContents` does not take much time.

Test file 1: ~350000 words, 1000 headings, 100 of them with markup:
Typing is rather slow, esp. due to word count and char count taking considerable amount of time.

I would say this is OK and with files this large, one would probably benefit from splitting them, at least for easier navigation if not for the speed.

Minor annoyances:
- text of the headings that have not changed is still processed every time the TOC is updated (but at least it is fast),
- I had to hardcode backtick and some HTML markup characters into `updateTableOfContents` in `store.ts` to process the headings without the need for `sanitizeHtml`.

I can go forward with a PR.

**Comment by nathanlesage** [comment]
Feel free to open a PR but I'm still convinced a simple memoize function would be the easiest way out, it shouldn't take up too much space. If we assume ~ 30 byte per heading then even with 10,000 different headings the cache would be only 300 kb, which should work super easy for any RAM size. Plus it would make your job considerably easier since you only add two or three lines of code. Something like that:

```javascript
function toc2html(text) {
  if (cache.has(text)) {
    return cache.get(text)
  } else {
    cache.set(text, text2html(text))
    return toc2html(text)
  }
}
```

**Comment by waterfallpit** [comment]
Yeah, you're right. It probably took me the same small amount of time to implement it now with cache as it did to search for how memoization is done in vue. I thought replacing `sanitizeHtml` with something simpler makes sense, but, oh well, it's important to learn, heh.

**Comment by jwhco** [comment]
Any progress? Keyboard input lag was addressed in https://github.com/Zettlr/Zettlr/issues/3295, and I raised it again in https://github.com/Zettlr/Zettlr/discussions/5601

There is a lot going on between keystrokes; if input were a priority, the tool would be more efficient for writing papers.

**Comment by ananthajithkr** [comment]
This issue seems to go away when I switch 'Save Modifications' to 'Off' position.

**Comment by ghost** [comment]
This is still an issue for me. For example: I have an ~15k word journal file in markdown; as I type in it daily, the lag grows longer and quickly becomes persistent. 
Closing and reopening Zettlr *slightly* improves things for a very brief time, then the lag returns. 
I also notice this with *any* longer file (> ~ 5-10k words).
NOTE: I have 'Save Modifications' set to OFF. Switching this option on/off seems to have no effect on the issue.

Is there some sort of...buffering/autosave issue occurring? Anything I can disable to help mitigate it?

Language: English
OS: Ubuntu 24.04
Zettlr version: 3.2.0
Framework 13, AMD Ryzen 5, 32GB RAM

**Comment by benniekiss** [comment]
Closing as both word counting and TOC rendering have seen significant performance improvements. If performance problems persist, please open a new report.
